Framework/DataHandling/src/ImggAggregateWavelengths.cpp

```diff
@@ -0,0 +1,834 @@
+ *
+ * @param bands number of output bands, or number of bands into
+ * which the input bands should be aggregated, non-overlapping and
+ * uniformly distributed.
+ *
+ * @param outDir where to write the output image(s).
+ *
+ * @param prefix prefix for the image names. The indices will be
+ * appended to the prefix
+ *
+ * @param outImgIndex index of the directory/angle/projection, for
+ * file naming purposes
+ */
+void ImggAggregateWavelengths::processDirectory(const Poco::Path &inDir,
+                                                size_t bands,
+                                                const std::string &outDir,
+                                                const std::string &prefix,
+                                                size_t outImgIndex) {
+  Mantid::API::MatrixWorkspace_sptr aggImg;
+
+  auto images = findInputImages(inDir);
+
+  if (images.empty()) {
+    g_log.warning()
+        << "Could not find any input image files in the subdirectory '"
+        << inDir.toString() << "'. It will be ignored." << std::endl;
+    return;
+  }
+
+  std::vector<std::pair<size_t, size_t>> ranges =
+      splitSizeIntoRanges(images.size(), bands);
+  processDirectory(inDir, ranges, outDir, prefix, outImgIndex);
+}
+
+/**
+ * Split into uniform blocks, for when producing multiple output bands
+ * from the input images/bands. If the division is not exact, a few
+ * images at the end will be ignored, so the number of input bands
+ * that will go into every output band is the same for all of
+ * them. For example, if 1000 input images/bands are aggregated into 3
+ * output bands, From 1000/3: 333 (output band 1), 333 (output band
+ * 2), 333 (output band 3), and 1 last image is not considered.
+ *
+ * @param availableCount how many input elements there are @param
+ * bands into how many uniform blocks they should be split
+ *
+ * @returns pairs of min-max for the output blocks
+ */
+std::vector<std::pair<size_t, size_t>>
+ImggAggregateWavelengths::splitSizeIntoRanges(size_t availableCount,
+                                              size_t bands) {
+  std::vector<std::pair<size_t, size_t>> ranges;
+  size_t inc = availableCount / bands;
+  if (inc < 1) {
+    throw std::runtime_error(
+        "The number of output bands requested (" + std::to_string(bands) +
+        ") is bigger than the number of available input images (" +
+        std::to_string(availableCount) + ". It should be equal or smaller, and "
+                                         "normally it is much smaller. Please "
+                                         "check that you are providing the "
+                                         "correct input parameters");
+  }
+
+  for (size_t count = 0; count < availableCount; count += inc) {
+    size_t max =
+        ((count + inc) > availableCount) ? availableCount : count + inc;
+    // TODO call the 'ranges' version with the convenient range
+    ranges.emplace_back(std::make_pair(count, max));
```

Comment:
With `emplace_back` you don't need to use `std::make_pair`. See, for example, #15874.